File: array/PrefixSum/python/PrefixSum.py
I64_MIN = -(1 << 63)
I64_MAX = (1 << 63) - 1


def _checked_add(left: int, right: int) -> int | None:
    total = left + right
    if total < I64_MIN or total > I64_MAX:
        return None
    return total


def _checked_sub(left: int, right: int) -> int | None:
    total = left - right
    if total < I64_MIN or total > I64_MAX:
        return None
    return total


def build_prefix_sums(values: list[int]) -> list[int] | None:
    prefix = [0] * (len(values) + 1)
    for index, value in enumerate(values):
        total = _checked_add(prefix[index], value)
        if total is None:
            return None
        prefix[index + 1] = total
    return prefix


def range_sum(prefix: list[int], left: int, right: int) -> int | None:
    return _checked_sub(prefix[right + 1], prefix[left])

File: array/PrefixSum/rust/src/prefix_sum.rs
pub fn build_prefix_sums(values: &[i64]) -> Option<Vec<i64>> {
    let mut prefix = vec![0_i64; values.len() + 1];
    for (index, value) in values.iter().enumerate() {
        prefix[index + 1] = prefix[index].checked_add(*value)?;
    }
    Some(prefix)
}

pub fn range_sum(prefix: &[i64], left: usize, right: usize) -> Option<i64> {
    prefix[right + 1].checked_sub(prefix[left])
}
